rebuild: add progress and map child states

tiagolobocastro opened the pull request:
Adds a new get_rebuild_progress rpc to indicate the rebuild progress
Also updates the list_nexus child field:
1. adds a rebuild progress to the child
2. maps the child states to a more simplified version for moac

CAS-181
CAS-193

tiagolobocastro committed:
rebuild: add progress and map child states

Adds a new get_rebuild_progress rpc to indicate the rebuild progress
Also updates the list_nexus child field:
1. adds a rebuild progress to the child
2. maps the child states to a more simplified version for moac

Also fixed a bug where a non faulted child could self rebuild itself
and added tests for this

CAS-181
CAS-193

paulyoong commented on mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs:
```diff
@@ -37,17 +37,24 @@ impl Nexus {
         {
-            Some(c) => Ok(c),
+            Some(c) if c.state == ChildState::Faulted => Ok(c),
+            Some(_) => Err(Error::ChildNotFaulted {
+                child: name.to_owned(),
+                name: self.name.clone(),
+            }),
```
Would it be useful to return the child state if not faulted?

tiagolobocastro replied:
yeah good idea, Done

paulyoong commented on mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs:
```diff
@@ -71,14 +78,29 @@ impl Nexus {
     }
 
+    /// Returns a `child` as an rpc Child type
+    pub fn to_child_rpc(&self, child: &NexusChild) -> Child {
```
Maybe the function name should be "to_rpc_child" which would match the comment.

tiagolobocastro replied:
Done

paulyoong commented on mayastor/src/bdev/nexus/nexus_child.rs:
```diff
@@ -100,6 +100,33 @@ impl ToString for ChildState {
+///     Rebuilding
+/// so here we inject a "rebuilding" state into the ChildState
+impl ChildState {
+    pub(crate) fn to_moac(self, rebuilding: bool) -> String {
```
Should we explicitly use the term "moac"? Maybe we should make the naming more generic in case moac is switched out for something else. Maybe something like "to_control_plane_state" or "to_external_state"?

tiagolobocastro replied:
yep we shouldn't... hmm to_rpc_state?

tiagolobocastro replied:
to_public

paulyoong commented on mayastor/src/replicas/rebuild/rebuild_impl.rs:
```diff
@@ -334,14 +337,39 @@ impl std::fmt::Display for RebuildOperation {
-            self.state(), self.source, self.destination,
-            self.start, self.end, self.next, self.block_size
+            "State: {}, Src: {}, Dst: {}, start: {}, end: {}, next: {}, \
+             block: {}, segment_sz: {}, recovered_blks: {}, progress: {}%",
```
block_size has been renamed to block. I think block_size was clearer.

tiagolobocastro replied:
oo don't know why I did that... fixed

paulyoong reviewed (dismissed):
lgtm

paulyoong reviewed (dismissed):
lgtm

tiagolobocastro committed:
Improve the state/status confusion

Avoid "converting" the Nexus and Child states into control plane
specific states by adding a new status for the Nexus and Child
The state now is an internal representation and the status is a higher
level view of the status (a bit like the health) which fits into what
the control plane currently requires.

the Nexus status is now inferred from the its children status:
State               Status
Init       ->       Degraded
Closed     ->       Faulted
Open       ->       All children online: Online
                    One or more child degraded/faulted: Degraded
                    All children faulted: Faulted
